File: lib/src/memory/runtime_memory_graph.dart
import '../crypto/kaalka_runtime.dart';
import '../graph/runtime_graph.dart';

Map<String, dynamic> buildRuntimeMemoryGraph(
  RuntimeGraph graph, [
  List<dynamic> history = const [],
]) {
  final normalized = normalizeRuntimeGraph(graph);
  final entities = normalized.nodes
      .map((n) => {
            'id': n.id,
            'type': n.type,
            'relations': normalized.edges
                .where((e) => e.source == n.id)
                .map((e) => e.target)
                .toList(),
          })
      .toList();
  final relations = normalized.edges
      .map((e) => {
            'from': e.source ?? e.from,
            'to': e.target ?? e.to,
            'type': e.type,
          })
      .toList();
  return {
    'entities': entities,
    'relations': relations,
    'graph_fingerprint':
        computeDeterministicHash({'entities': entities, 'relations': relations, 'history_len': history.length}),
    'bounded': true,
  };
}

File: src/memory/runtimeMemoryGraph.ts
import {
  RuntimeGraphContract,
  type RuntimeGraph,
  type RuntimeNode,
  type RuntimeEdge,
} from "../contracts/graphContracts.js";
import { computeKaalkaHashPayload } from "../crypto/kaalkaRuntime.js";

export type MemoryGraphEntity = { id: string; type: string; relations: string[] };
export type RuntimeMemoryGraph = {
  entities: MemoryGraphEntity[];
  relations: Array<{ from: string; to: string; type: string }>;
  graph_fingerprint: string;
  bounded: boolean;
};

export function buildRuntimeMemoryGraph(
  graph: RuntimeGraph,
  history: unknown[] = [],
): RuntimeMemoryGraph {
  const normalized = RuntimeGraphContract.normalize(graph) as {
    nodes: RuntimeNode[];
    edges: RuntimeEdge[];
  };
  const entities: MemoryGraphEntity[] = normalized.nodes.map((n: RuntimeNode) => ({
    id: String(n.id),
    type: String(n.type ?? "node"),
    relations: normalized.edges
      .filter((e: RuntimeEdge) => e.source === n.id || e.from === n.id)
      .map((e: RuntimeEdge) => String(e.target ?? e.to)),
  }));
  const relations = normalized.edges.map((e: RuntimeEdge) => ({
    from: String(e.source ?? e.from),
    to: String(e.target ?? e.to),
    type: String(e.type ?? "link"),
  }));
  return {
    entities,
    relations,
    graph_fingerprint: computeKaalkaHashPayload({ entities, relations, history_len: history.length }),
    bounded: true,
  };
}
